.github/workflows/ci-pipeline.yml

```diff
@@ -8,10 +8,47 @@ on:
       - "**/README.md"
 
 jobs:
-  build:
-    name: Build and test
+  CI_Build_And_Test:
+    name: CI Build & Test
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
-    - run: docker build --target test --tag todo-app:test .
-    - run: docker run todo-app:test
\ No newline at end of file
+
+    - name: Docker build
+      run: docker build --target test --tag todo-app:test .
+
+    - name: Docker run tests
+      run: docker run todo-app:test
+
+  CD_Publish_Docker_Hub:
+    name: CD Publish to Docker Hub
+    needs: [CI_Build_And_Test]
+    if: ${{ (github.ref == 'refs/heads/dev/Module8-ProjectExercise' || github.ref == 'refs/heads/main') && github.event_name == 'push' }}
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    
+    - name: Login to Docker Hub
+      run: docker login --username ${{ secrets.DOCKER_HUB_USERNAME }} --password ${{ secrets.DOCKER_HUB_PASSWORD }}
```

Comment:
We should ideally action this step using the `--password-stdin` flag rather than passing the password explicitly, to avoid leaking those credentials to shell logs. [See the docs](https://docs.docker.com/engine/reference/commandline/login/) for more info, example of how that might look below:
```
echo ${{ secrets.DOCKER_HUB_PASSWORD }} | docker login --username ${{ secrets.DOCKER_HUB_USERNAME }} --password-stdin
```